zha/application/platforms/sensor/__init__.py

```diff
@@ -264,14 +270,27 @@ def handle_cluster_handler_attribute_updated(
         ):
             self.maybe_emit_state_changed_event()
 
-    def formatter(self, value: int | enum.IntEnum) -> int | float | str | None:
+    def formatter(
+        self, value: int | enum.IntEnum
+    ) -> date | datetime | int | float | str | None:
+        """Pass-through formatter."""
+        return getattr(
+            self,
+            NON_NUMERIC_FORMATTERS.get(self._attr_device_class, "numeric_formatter"),
```

Comment:
I think it would be clearest to have a single `formatter` function for the base sensor (`converter`?) and then have each sub-type override it, having `NumericSensor`, `StringSensor`, and `DateTimeSensor` provide an implementation.